frontend/src/components/widgets/buttons/FloatingMenuToggleButton.vue

```diff
@@ -0,0 +1,70 @@
+<template>
+	<div class="toggle-button" @mousedown.stop="toggleFloatingMenu()" @mouseenter="maybeOpenOnHover()">
```

Comment:
I'm concerned that we are using a wrapper div for our click and hover events. This is, for example, triggering outside the rounded corners of the widget. It's much nicer if we can trigger on the HTML `<button>` element itself for widgets which use that HTML element, which is what the `:action="..."` prop is for in many button/input widgets. I wonder if we can alter this approach a little in order to make the button/input widget, which is supposed to open the floating menu, pipe its `action` prop callback as the trigger. I also find that the extra double wrappers required for the two template tags inside this component's tag makes the code more complex and visually noisy. Since these are, in my opinion, a pretty big compromise and I am not sure the previous approach without this `FloatingMenuToggleButton` component was any worse, it might be worth exploring other options or chatting about whether it's worth reverting this particular component's changes back to how it was before and then finding a simpler way to clean it up. I'd love to discuss the pros and cons of each, or see if we might be able to find a cleaner way to do it, over Discord if you have any thoughts.

Reply:
Yeah I take your point. I really wanted to use a Directive for this but unfortunately Vue Custom Directives are quite limited in what they can do. What this component provides is the logic to navigate between group items by hovering. That logic is needed in enough places that it seemed worthwhile to implement it once instead of having a bunch of repeated code. I'm imagining that this component would also expand in the future to help with keyboard nav and a11y, again cutting down on repeated code. And I've been thinking about logic to allow menu items to be selected with one click-and-drag gesture (eg. to change blend modes mouse down on the toggle button, drag to the one you want, and mouse up to select it).

The specific bug about rounded corners and that sort of thing could be addressed with an extra check in the event handlers (good catch on that one). Again, a bit overcomplicated, but at least that's just internal implementation details.

No argument that the template code is a bit overly verbose. Still, in my opinion it is worth it to have one component that handles all our various floating menu toggles and plays nice with everything else.

Reply:
What if we take the following approach instead of relying on a wrapper component?:

We create a TS file in utilities that acts as a floating menu manager (this is what I described above about moving the `provide/inject` logic into a utilities TS file). In every component where we have a button (or any input widget with an `:action="..."` prop) we import a function from that utility TS file and feed it to the `action` prop. For example, if the function is called `openAssociatedFloatingMenu`, a button input component could be given the prop `:action="openAssociatedFloatingMenu('some-unique-id-for-the-associated-popover')"`. Then the FloatingMenu component can be given a required prop like `:id="'some-unique-id-for-the-associated-popover'"` which is used by the FloatingMenu to register itself with the utility TS file. In addition to `:action" props, we could also add a ":hoverAction" prop to deal with the hover.

I don't know if the above description is the best, but it is one possible way to get around using a wrapper component which I still have reservations about because of multiple compromises it requires. I'd love to chat on Discord about other ways we may be able to reduce the code and minimize the complexity. I'm just concerned that the current approach in this part of the PR adds both code and complexity, even though I understand your goal is to generalize things as more components are added that need to utilize common functionality. There is probably a better way neither of us have thought about yet, we should discuss the best option. (Or perhaps break this out from the rest of the PR and get the other parts merged now as we defer on this topic specifically.)

Reply:
That's an interesting idea, but I think in practice more cumbersome than the current approach.  one thing it needs `:action` but also mouseenter and eventually mousedrag. For another, I don't think there's a way to directly pass injected properties into a template, so each component that includes a menu button would need to add those 2-3 functions to its local state.

This does seem like the perfect use case for a Directive, and with some more Googling I think it might be possible (although a bit hacky) with Vue custom directives. That directive could be applied to an existing button element like `<my-button v-floating-menu-trigger='"some-unique-id"'>`, and would apply all the necessary event handlers internally. Adding native event handlers is easy enough, although I'm not 100% sure there's a way to set the :action handler on the button.

Personally though I'd miss having the buttons and the menus tied together inside one component, I think it's just a little bit more idiot-proof and saves us from having to think about the case where multiple buttons toggle the same menu id.